app/src/main/java/com/github/swent/echo/data/room/RoomLocalDataSource.kt

```diff
@@ -42,20 +43,54 @@ class RoomLocalDataSource @Inject constructor(db: AppDatabase) : LocalDataSource
     }
 
     override suspend fun setAssociation(association: Association) {
+        val crossRefs =
+            association.relatedTags.map {
+                AssociationTagCrossRef(association.associationId, it.tagId)
+            }
+
+        associationDao.deleteAllAssociationTagCrossRefsForAssociation(association.associationId)
+        tagDao.insertAll(association.relatedTags.toTagRoomList())
         associationDao.insert(AssociationRoom(association))
+        associationDao.insertAssociationTagCrossRefs(crossRefs)
+    }
+
+    override suspend fun getAssociations(
+        associationIds: List<String>,
+        syncedSecondsAgo: Long
+    ): List<Association> {
+        val after = computeTimestamp(syncedSecondsAgo)
+        return associationDao.get(associationIds, after).toAssociations()
     }
 
     override suspend fun getAllAssociations(syncedSecondsAgo: Long): List<Association> {
         val after = computeTimestamp(syncedSecondsAgo)
         return associationDao.getAll(after).map { it.toAssociation() }
     }
 
+    override suspend fun getAssociationIds(
+        associationIds: List<String>,
+        syncedSecondsAgo: Long
+    ): List<String> {
+        return associationDao.getIdsFrom(associationIds, computeTimestamp(syncedSecondsAgo))
+    }
+
     override suspend fun getAllAssociationIds(secondsAgo: Long): List<String> {
         return associationDao.getAllIds(computeTimestamp(secondsAgo))
     }
 
     override suspend fun setAssociations(associations: List<Association>) {
+        val tags = associations.flatMap { it.relatedTags }.toSet()
+        val crossRefs =
+            associations.flatMap { association ->
+                association.relatedTags.map {
+                    AssociationTagCrossRef(association.associationId, it.tagId)
+                }
+            }
+
+        eventDao.deleteAllEventTagCrossRefsForEvents(associations.map { it.associationId })
```

Comment:
I think this should be `deleteAllAssociationTagCrossRefsForAssociation`.

Reply:
nice catch! thanks!
you're absolutely right! I made a copy paste error here :see_no_evil: